sympy/solvers/solveset.py

```diff
@@ -635,11 +634,12 @@ def _solveset(f, symbol, domain, _check=False):
             result = FiniteSet(*[s for s in result
                       if isinstance(s, RootOf)
                       or domain_check(fx, symbol, s)])
-
+    if _was_domain:
+        return Intersection(result, domain)
     return result
```

Comment:
You've identified the problem as being the lack of domain handling in `solve_real_trig`. I recommend doing

```diff
diff --git a/sympy/solvers/solveset.py b/sympy/solvers/solveset.py
index 9f413e4..7dd3d92 100644
--- a/sympy/solvers/solveset.py
+++ b/sympy/solvers/solveset.py
@@ -363,7 +363,7 @@ def _solve_as_rational(f, symbol, domain):
         return valid_solns - invalid_solns


-def _solve_real_trig(f, symbol):
+def _solve_trig(f, symbol, domain):
     """ Helper to solve trigonometric equations """
     f = trigsimp(f)
     f_original = f
@@ -379,8 +379,9 @@ def _solve_real_trig(f, symbol):
     solns = solveset_complex(g, y) - solveset_complex(h, y)

     if isinstance(solns, FiniteSet):
-        return Union(*[invert_complex(exp(I*symbol), s, symbol)[1]
-                       for s in solns])
+        result = Union(*[invert_complex(exp(I*symbol), s, symbol)[1]
+            for s in solns])
+        return Intersection(result, domain)
     elif solns is S.EmptySet:
         return S.EmptySet
     else:
@@ -574,7 +575,7 @@ def _solveset(f, symbol, domain, _check=False):
         result = Union(*[solver(m, symbol) for m in f.args])
     elif _is_function_class_equation(TrigonometricFunction, f, symbol) or \
             _is_function_class_equation(HyperbolicFunction, f, symbol):
-        result = _solve_real_trig(f, symbol)
+        result = _solve_trig(f, symbol, domain)
     elif f.is_Piecewise:
         dom = domain
         result = EmptySet()
diff --git a/sympy/solvers/tests/test_solveset.py b/sympy/solvers/tests/test_solveset.py
index 7af972f..83fc70b 100644
--- a/sympy/solvers/tests/test_solveset.py
+++ b/sympy/solvers/tests/test_solveset.py
@@ -737,9 +737,7 @@ def test_solve_trig():

     y, a = symbols('y,a')
     assert solveset(sin(y + a) - sin(y), a, domain=S.Reals) == \
-        Union(imageset(Lambda(n, 2*n*pi), S.Integers),
-        imageset(Lambda(n,
-        -I*(I*(2*n*pi +arg(-exp(-2*I*y))) + 2*im(y))), S.Integers))
+        imageset(Lambda(n, 2*n*pi), S.Integers)
```

That name change will make it more clear that the function is not solving for a real solution.